p2p/ipld/plugin/plugin.go

```diff
@@ -0,0 +1,328 @@
+	return nil
+}
+
+func (n nmtNode) Resolve(path []string) (interface{}, []string, error) {
+	switch path[0] {
+	case "0":
+		return &node.Link{Cid: cidFromNamespacedSha256(n.l)}, path[1:], nil
+	case "1":
+		return &node.Link{Cid: cidFromNamespacedSha256(n.r)}, path[1:], nil
+	default:
+		return nil, nil, errors.New("invalid path for inner node")
+	}
+}
+
+func (n nmtNode) Tree(path string, depth int) []string {
+	if path != "" || depth == 0 {
+		return nil
+	}
+
+	return []string{
+		"0",
+		"1",
+	}
+}
+
+func (n nmtNode) ResolveLink(path []string) (*node.Link, []string, error) {
+	obj, rest, err := n.Resolve(path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	lnk, ok := obj.(*node.Link)
+	if !ok {
+		return nil, nil, fmt.Errorf("was not a link")
+	}
+
+	return lnk, rest, nil
+}
+
+func (n nmtNode) Copy() node.Node {
+	panic("implement me")
+}
+
+func (n nmtNode) Links() []*node.Link {
+	return []*node.Link{{Cid: cidFromNamespacedSha256(n.l)}, {Cid: cidFromNamespacedSha256(n.r)}}
+}
+
+func (n nmtNode) Stat() (*node.NodeStat, error) {
+	return &node.NodeStat{}, nil
+}
+
+func (n nmtNode) Size() (uint64, error) {
+	return 0, nil
+}
+
+type nmtLeafNode struct {
+	cid  cid.Cid
+	Data []byte
+}
+
+func (l nmtLeafNode) RawData() []byte {
+	return append([]byte{nmt.LeafPrefix}, l.Data...)
+}
+
+func (l nmtLeafNode) Cid() cid.Cid {
+	return l.cid
+}
+
+func (l nmtLeafNode) String() string {
+	return fmt.Sprintf(`
+leaf {
+	hash: 		%x,
+	len(Data): 	%v
+}`, l.cid.Hash(), len(l.Data))
+}
+
+func (l nmtLeafNode) Loggable() map[string]interface{} {
+	return nil
+}
+
+func (l nmtLeafNode) Resolve(path []string) (interface{}, []string, error) {
+	if path[0] == "Data" {
+		// TODO: store the data separately
+		// currently nmtLeafNode{Data:} contains the actual data
+		// instead, there should be a link in the leaf to the actual data
+		return nil, nil, nil
+	} else {
+		return nil, nil, errors.New("invalid path for leaf node")
+	}
+}
+
+func (l nmtLeafNode) Tree(path string, depth int) []string {
+	if path != "" || depth == 0 {
+		return nil
+	}
+
+	return []string{
+		"Data",
+	}
+}
+
+func (l nmtLeafNode) ResolveLink(path []string) (*node.Link, []string, error) {
+	obj, rest, err := l.Resolve(path)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	lnk, ok := obj.(*node.Link)
+	if !ok {
+		return nil, nil, fmt.Errorf("was not a link")
+	}
+	return lnk, rest, nil
+}
+
+func (l nmtLeafNode) Copy() node.Node {
+	panic("implement me")
+}
+
+func (l nmtLeafNode) Links() []*node.Link {
+	return []*node.Link{{Cid: l.Cid()}}
+}
+
+func (l nmtLeafNode) Stat() (*node.NodeStat, error) {
+	return &node.NodeStat{}, nil
+}
+
+func (l nmtLeafNode) Size() (uint64, error) {
+	return 0, nil
+}
+
+func cidFromNamespacedSha256(namespacedHash []byte) cid.Cid {
+	if got, want := len(namespacedHash), 2*namespaceSize+sha256.Size; got != want {
+		panic(fmt.Sprintf("expected a namespace prefixed sha256, got: %v, want: %v", got, want))
```

Comment:
Just noticed: This should not panic in all cases. So it's better to bubble up the error!